src/Butler/Auth/OIDC.hs

```diff
@@ -0,0 +1,206 @@
+                            (oidc oidcEnv)
+                            (manager oidcEnv)
+                            (encodeUtf8 oauthState)
+                            (encodeUtf8 oauthCode)
+                setSession (UserName tokens.idToken.sub)
+
+    guestCallbackRoute :: Servant.Handler AuthResp
+    guestCallbackRoute = setSession "guest"
+
+    setSession :: UserName -> Servant.Handler AuthResp
+    setSession username = do
+        session <- liftIO $ runProcessIO os process $ newSession sessions username
+        liftIO (SAS.acceptLogin cookieSettings jwtSettings session.sessionID) >>= \case
+            Just r -> do
+                let page = workspaceUrl mWorkspace
+                pure $ r (script_ $ "window.location.href = " <> showT page)
+            Nothing -> error "oops?!"
+
+    indexRoute :: AuthResult SessionID -> Servant.Handler (Html ())
+    indexRoute = \case
+        Authenticated sessionID -> do
+            mSession <- atomically (lookupSession sessions sessionID)
+            case mSession of
+                Just _ ->
+                    pure $ mkIndexHtml (websocketHtml (workspaceUrl mWorkspace) sessionID)
+                Nothing -> do
+                    liftIO $ runProcessIO os process $ logError "no more auth?" ["id" .= sessionID]
+                    pure "Unknown session"
+        _ -> throwError $ err303{errHeaders = [("Location", encodeUtf8 (workspaceUrl mWorkspace) <> "_guest_cb")]}
+
+    loginRoute :: OIDCEnv -> Servant.Handler NoContent
+    loginRoute oidcEnv = do
+        redirectLocation <- liftIO genOIDCURL
+        throwError $ err303{errHeaders = [("Location", redirectLocation)]}
+      where
+        genOIDCURL :: IO ByteString
+        genOIDCURL = do
+            loc <-
+                O.prepareAuthenticationRequestUrl
+                    (mkSessionStore oidcEnv Nothing Nothing)
+                    oidcEnv.oidc
+                    [O.openId]
+                    mempty
+            pure . B.pack $ show loc
+
+cookieSettings :: CookieSettings
+cookieSettings =
+    defaultCookieSettings
+        { cookieSameSite = SameSiteStrict
+        , cookieXsrfSetting = Nothing
+        , cookieExpires = Just (UTCTime (fromGregorian 2030 1 1) 0)
+        }
+
+oIDCAuthApp :: Sessions -> (Html () -> Html ()) -> ProcessIO AuthApplication
+oIDCAuthApp sessions mkIndexHtml = do
+    JwkStorage myKey <- fst <$> newProcessMemory "display-key.jwk" (JwkStorage <$> liftIO generateKey)
+    let jwtSettings = defaultJWTSettings myKey
+    let cfg = cookieSettings :. jwtSettings :. EmptyContext
+    env <- ask
+    Just client_id <- liftIO $ getEnv "OIDC_ID"
+    Just client_password <- liftIO $ getEnv "OIDC_PASSWORD"
+    oidcenv <-
+        liftIO . initOIDCEnv $
+            OIDCProviderConfig
+                "https://accounts.google.com"
+                (decodeUtf8 client_id)
+                (decodeUtf8 client_password)
+                "https://localhost:8080/"
```

Comment:
Not sure where this should be stored, but AuthApplication and AppContext should have accessed to the base PUBLIC_URL. I guess this should be part of the Display data, and it would be provided to AuthApplication through DisplayApplication. So for now, this should also be a parameter of 'oIDCAuthApp'.

Reply:
Sure I'll propose a followup with this.

Reply:
Probably this publicURL must be passed as an environment variable ?

Reply:
How about we do the following:
- Add a hostname to the 'DisplayAddr' data
- Store the DisplayAddr in the 'Display' data (so that it is available to App through the AppContext)
- Pass the DisplayAddr to the AuthApplication
- Provide a "DisplayAddr -> Text" function to render the url, e.g. into https://localhost:8080.

Reply:
Nevermind about this suggestion, let's construct the redirection url using the Host header, similar to what is done in the Network module with serverName.